drivers/i3c/i3c_renesas_imx3112.c

```diff
@@ -0,0 +1,416 @@
+/* renesas_imx3112_i3c.c - I3C file for Renesas I3C 1:2 bus multiplexer */
+
+/*
+ * Copyright (c) 2022 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "i3c_renesas_imx3112_regs.h"
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i3c.h>
+#include <zephyr/drivers/i3c/ccc.h>
+#include <zephyr/drivers/i3c/devicetree.h>
+#include <zephyr/kernel.h>
+#include <zephyr/types.h>
+
+#define LOG_LEVEL CONFIG_I3C_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(i3c_renesas_imx3112);
+
+/* Number of ports that can be accessed by device */
+#define NUM_CHANNELS 2
+/* Timeout for locking in a channel - 5 seconds is more than sufficient */
+#define CHAN_LOCK_TIMEOUT K_MSEC(5000)
+
+/*
+ * Resistor values for SCL/SDA lines.
+ * Not used if has_external_resistor is true
+ */
+struct channel_resistance {
+	uint8_t scl;
+	uint8_t sda;
+};
+
+/*
+ * Devicetree values used to configure the mux during initalization.
+ * Determine the electrical characteristics of the mux channels
+ */
+struct mux_hw_configuration {
+	/*
+	 * Indicates if mux has external pullup resistors or
+	 * internal programmable resistors for SCL/SDA lines
+	 */
+	bool has_external_resistor;
+	/* IO voltage level - at which SCL/SDA lines are driven */
+	uint8_t io_voltage;
+	/* Channel resistances */
+	struct channel_resistance ch_resistance[NUM_CHANNELS];
+};
+
+/*
+ * I3C device configuration at initialization phase
+ *
+ * These are mostly configured with device tree information.
+ */
+struct i3c_renesas_imx3112_mux_config {
+	/* I3C bus controller */
+	const struct device *bus;
+
+	/* Device id of the mux in the  bus */
+	const struct i3c_device_id dev_id;
+
+	/* Hw config for the mux */
+	struct mux_hw_configuration hw_config;
+};
+
+/*
+ * I3C device data updated at runtime
+ */
+struct i3c_renesas_imx3112_mux_data {
+	struct {
```

Comment:
Does this have to be a struct? Is this getting re-used somewhere?

Reply:
Which struct do you refer to, mux_data or mux_state?

Reply:
If you're talking about the mutex state, i'm just separating the lock functionality from everything else for simplicity.